bookstore/handlers.py

```diff
@@ -19,13 +25,104 @@ def get(self):
         self.finish(json.dumps({"bookstore": True, "version": version}))
 
 
+# NOTE: We need to ensure that publishing is not configured if bookstore settings are not
+# set. Because of how the APIHandlers cannot be configurable, all we can do is reach into settings
+
+
+class BookstorePublishHandler(APIHandler):
+    """Publish a notebook to the publish path"""
+
+    def __init__(self, *args, **kwargs):
+        super(APIHandler, self).__init__(*args, **kwargs)
+        # create an easy helper to get at our bookstore settings quickly
+        self.bookstore_settings = BookstoreSettings(
+            config=self.settings['config']['BookstoreSettings']
+        )
+
+        self.fs = s3fs.S3FileSystem(
+            key=self.bookstore_settings.s3_access_key_id,
+            secret=self.bookstore_settings.s3_secret_access_key,
+            client_kwargs={
+                "endpoint_url": self.bookstore_settings.s3_endpoint_url,
+                "region_name": self.bookstore_settings.s3_region_name,
+            },
+            config_kwargs={},
+            s3_additional_kwargs={},
+        )
+
+    @property
+    def bucket(self):
+        return self.bookstore_settings.s3_bucket
+
+    @property
+    def prefix(self):
+        return self.bookstore_settings.published_prefix
+
+    def s3_path(self, path):
+        """compute the s3 path based on the bucket, prefix, and the path to the notebook"""
+        return s3_path(self.bucket, self.prefix, path)
+
+    def _publish(self, model, path):
+        if model['type'] != 'notebook':
+            raise web.HTTPError(400, "bookstore only publishes notebooks")
+        content = model['content']
+
+        full_s3_path = self.s3_path(path)
+
+        self.log.info("Publishing to %s", s3_display_path(self.bucket, self.prefix, path))
+
+        # Likely implementation:
+        #
+        # with self.fs.open(full_s3_path, mode="wb") as f:
+        #     f.write(content.encode("utf-8"))
```

Comment:
Now that it's been some time, I've played with the [`aiobotocore`](https://github.com/aio-libs/aiobotocore) implementation for S3 again and it's... pretty good actually. Appears to work with the use cases we have in mind, complete with allowing for using IAM roles in automation (no having to specify keys):

```python
In [9]: import asyncio

In [10]: import aiobotocore

In [11]: session = aiobotocore.get_session()

In [13]: client = session.create_client('s3')

In [14]: client
Out[14]: <aiobotocore.client.S3 at 0x7fc2f7557550>

In [15]: obj = await client.get_object(Bucket="super-cool-notebooks", Key="workspace/kylek/titles.ipynb")

In [16]: obj
Out[16]:
{'ResponseMetadata': {'RequestId': '47BCFC8BA2C31223',
  'HostId': 'ShBbToxlM+4NHx6UcstZI4kveRZRIwcI5AdaR83fgRQMXjr9bUhhWazgz4mS77Pi/307sIIMzvc=',
  'HTTPStatusCode': 200,
  'HTTPHeaders': {'x-amz-id-2': 'ShBbToxlM+4NHx6UcstZI4kveRZRIwcI5AdaR83fgRQMXjr9bUhhWazgz4mS77Pi/307sIIMzvc=',
   'x-amz-request-id': '47BCFC8BA2C31223',
   'date': 'Wed, 23 Jan 2019 23:42:33 GMT',
   'last-modified': 'Mon, 03 Dec 2018 20:12:34 GMT',
   'etag': '"ceaa8fc15ca50afabc93d111892bc1be"',
   'x-amz-version-id': 'qvr.OG90WctPVd_somfr4knR3hqTXzOt',
   'accept-ranges': 'bytes',
   'content-type': 'binary/octet-stream',
   'content-length': '30756',
   'server': 'AmazonS3'},
  'RetryAttempts': 0},
 'AcceptRanges': 'bytes',
 'LastModified': datetime.datetime(2018, 12, 3, 20, 12, 34, tzinfo=tzutc()),
 'ContentLength': 30756,
 'ETag': '"ceaa8fc15ca50afabc93d111892bc1be"',
 'VersionId': 'qvr.OG90WctPVd_somfr4knR3hqTXzOt',
 'ContentType': 'binary/octet-stream',
 'Metadata': {},
 'Body': <StreamingBody at 0x7fc2f6c477c8 for ClientResponseContentProxy at 0x7fc31f459f48>}

```